arangod/IResearch/IResearchFilterFactory.cpp

```diff
@@ -1290,24 +1298,116 @@ std::pair<Result, aql::AstNodeType> buildBinaryArrayComparisonPreFilter(
             TRI_ASSERT(false);  // new array comparison operator?
             return {Result{TRI_ERROR_NOT_IMPLEMENTED,
                            "Unknown Array AT LEAST comparison operator"},
-                    aql::AstNodeType::NODE_TYPE_ROOT};
+                    aql::AstNodeType::NODE_TYPE_ROOT, 0};
         }
         break;
       default:
         TRI_ASSERT(false);  // new qualifier added ?
         return {Result{TRI_ERROR_NOT_IMPLEMENTED,
                        "Unknown qualifier in Array comparison operator"},
-                aql::AstNodeType::NODE_TYPE_ROOT};
+                aql::AstNodeType::NODE_TYPE_ROOT, 0};
     }
   }
-  return std::make_pair(TRI_ERROR_NO_ERROR, expansionNodeType);
+  return {TRI_ERROR_NO_ERROR, expansionNodeType, atLeastCount};
+}
+
+struct ByTermsOptimizationContext {
+  void Finalize() noexcept {
+    if (allMatch) {
+      for (auto& f : filters) {
+        TRI_ASSERT(f.second);
+        auto* opt = f.second->mutable_options();
+        opt->min_match = opt->terms.size();
+      }
+    }
+  }
+
+  containers::FlatHashMap<std::string, irs::by_terms*> filters;
+  bool useByTerms{false};
+  bool allMatch{false};
+};
+
+Result appendByTermsFilter(irs::boolean_filter* filter,
+                           ScopedAqlValue const& value, std::string& name,
+                           FilterContext const& filterCtx,
+                           ByTermsOptimizationContext& termsOpt) {
+  auto makeFilter = [&]() -> irs::by_terms* {
```

Comment:
Why don't return reference? In such case don't need to think about nullptr possibility on caller side